main-board/src/sd_interface.rs

```diff
@@ -0,0 +1,152 @@
+
+use core::{marker::PhantomData};
+
+use defmt::{warn, info};
+
+use crate::{
+    hal,
+    pac,
+    sd,
+};
+
+pub struct TimeSink {
+    _marker: PhantomData<*const ()>,
+}
+
+impl TimeSink { 
+    fn new() -> Self {
+        TimeSink {
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl sd::TimeSource for TimeSink {
+    fn get_timestamp(&self) -> sd::Timestamp {
+        sd::Timestamp {
+            year_since_1970: 0,
+            zero_indexed_month: 0,
+            zero_indexed_day: 0,
+            hours: 0,
+            minutes: 0,
+            seconds: 0,
+        }
+    }
+}
+
+pub struct SdInterface
+{
+    pub sd_controller: sd::Controller<sd::SdMmcSpi<hal::sercom::spi::Spi<hal::sercom::spi::Config<hal::sercom::spi::Pads<hal::sercom::Sercom1, hal::sercom::IoSet1, hal::gpio::Pin<hal::gpio::PA19, hal::gpio::Alternate<hal::gpio::C>>, hal::gpio::Pin<hal::gpio::PA16, hal::gpio::Alternate<hal::gpio::C>>, hal::gpio::Pin<hal::gpio::PA17, hal::gpio::Alternate<hal::gpio::C>>>>, hal::sercom::spi::Duplex>, hal::gpio::Pin<hal::gpio::PA18, hal::gpio::Output<hal::gpio::PushPull>>>, TimeSink>,
+    pub volume: sd::Volume,
+    pub root_directory: sd::Directory,
+    pub file: sd::File,
+
+}
+
+impl SdInterface {
+    pub fn new(mclk: &pac::MCLK, sercom: pac::SERCOM1, spi_clk: hal::clock::Sercom1CoreClock, cs: hal::gpio::Pin<hal::gpio::PA18, hal::gpio::Output<hal::gpio::PushPull>>, sck: hal::gpio::Pin<hal::gpio::PA17, hal::gpio::Output<hal::gpio::PushPull>>, miso: hal::gpio::Pin<hal::gpio::PA19, hal::gpio::Output<hal::gpio::PushPull>>, mosi: hal::gpio::Pin<hal::gpio::PA16, hal::gpio::Output<hal::gpio::PushPull>>) -> Self {
```

Comment:
Likewise, here (at the very least) we should have Pin<>, PAxx and PushPull imported so that this type is simplified. I think making it properly generic isn't really needed given that the the SD card will only ever be on one set of pins, but it might be nice to look at doing.

Reply:
Fair, making it generic is maybe overkill. We could maybe first use a type alias, and only make it generic in the future if we feel the need?

Reply:
Ok, so thinking more about it, I agree, no need for generics (unless they are trivial and improve readability and usability). For now, `SdInterface` should just encapsulate all the init and writing logic for the SD Card, and no need to make it general.

Reply:
@NoahSprenger, therefore, could you remove the SPI pins (sck, miso, mosi) in this constructor, and just hard-code them?